backend/loop.ts

```diff
@@ -0,0 +1,300 @@
+import { Executor, transact } from "./db";
+import {
+  setClientRecord,
+  ClientRecord,
+  mustGetClientRecords,
+  getCookie,
+} from "./data";
+import { Mutation } from "../schemas/push";
+import { ClientID, ClientMap } from "./server";
+import { sendPokes, ClientPokeResponse, computePokes } from "./poke";
+import { EntryCache } from "./entry-cache";
+import { ReplicacheTransaction } from "./replicache-transaction";
+import { PostgresStorage } from "./postgres-storage";
+
+export type Now = typeof performance.now;
+export type Sleep = (ms: number) => Promise<void>;
+export type ClientMutation = Required<Mutation> & { clientID: ClientID };
+export type RoomID = string;
+
+// Returns false when there is no more work to do.
+export type Step = () => Promise<void>;
+
+/**
+ * A game loop that runs some `Step` function periodically until the step has
+ * no more work to do.
+ */
+export class Loop {
+  private _step: Step;
+  private _now: Now;
+  private _sleep: Sleep;
+  private _loopIntervalMs: number;
+  private _running: boolean;
+  private _runPending: boolean;
+
+  constructor(step: Step, now: Now, sleep: Sleep, loopIntervalMs: number) {
+    this._step = step;
+    this._now = now;
+    this._sleep = sleep;
+    this._loopIntervalMs = loopIntervalMs;
+    this._running = false;
+    this._runPending = false;
+  }
+
+  async run() {
+    if (this._running) {
+      this._runPending = true;
+      return;
+    }
+
+    this._running = true;
+    try {
+      for (;;) {
+        const t0 = this._now();
+        this._runPending = false;
+        await this._step();
+        const t1 = this._now();
+        const elapsed = t1 - t0;
+        if (!this._runPending) {
+          break;
+        }
+        await this._sleep(Math.max(0, this._loopIntervalMs - elapsed));
+      }
+    } finally {
+      this._running = false;
+    }
+  }
+}
+
+/**
+ * Advances the game loop one step by running all mutations in all rooms that
+ * are ready and sending out the relevant pokes on the coresponding Clients.
+ *
+ * @param clients All currently connected clients.
+ * @param mutators All known mutators.
+ */
+export async function step(
+  clients: ClientMap,
+  mutators: Record<string, Function>
+): Promise<void> {
+  const t0 = Date.now();
+
+  const mutationsByRoom = getPendingMutationsByRoom(clients);
+  if (mutationsByRoom.size === 0) {
+    return;
+  }
+
+  const pokes = await transact(async (executor) => {
+    const pokes = [];
+    // TODO: We could parallelize this.
+    for (const [roomID, mutations] of mutationsByRoom) {
+      pokes.push(
+        ...(await stepRoom(
+          executor,
+          roomID,
+          mutations,
+          mutators,
+          [...clients.values()]
```

Comment:
Yeah this part was a bit of a mess in this version of the code. What I have moved to since is that conceptually a snapshot/copy of the client state is sent as a param into `step()`.